base/small_set.hpp

```diff
@@ -0,0 +1,165 @@
+#pragma once
+
+#include "base/assert.hpp"
+#include "base/bits.hpp"
+#include "base/macros.hpp"
+
+#include <cstdint>
+#include <iterator>
+#include <sstream>
+#include <string>
+
+namespace base
+{
+// A set of values less than |UpperBound|.
+//
+// Requires UpperBound + O(1) bits of memory.  All operations except
+// Clear() and iteration are O(1).  Clear() and iteration require
+// O(UpperBound) steps.
+//
+// *NOTE* This class *IS NOT* thread safe.
+template <uint64_t UpperBound>
+class SmallSet
+{
+public:
+  static uint64_t constexpr kOne = 1;
+  static uint64_t constexpr kNumBlocks = (UpperBound + 63) / 64;
+
+  class Iterator
+  {
+  public:
+    using difference_type = uint64_t;
+    using value_type = uint64_t;
+    using pointer = void;
+    using reference = void;
+    using iterator_category = std::forward_iterator_tag;
+
+    Iterator(uint64_t const * blocks, uint64_t current_block_index)
+      : m_blocks(blocks), m_current_block_index(current_block_index), m_current_block(0)
+    {
+      ASSERT_LESS_OR_EQUAL(current_block_index, kNumBlocks, ());
+      if (current_block_index < kNumBlocks)
+        m_current_block = m_blocks[current_block_index];
+      SkipZeroes();
+    }
+
+    bool operator==(Iterator const & rhs) const
+    {
+      return m_blocks == rhs.m_blocks && m_current_block_index == rhs.m_current_block_index &&
+             m_current_block == rhs.m_current_block;
+    }
+
+    bool operator!=(Iterator const & rhs) const { return !(*this == rhs); }
+
+    uint64_t operator*() const
+    {
+      ASSERT_NOT_EQUAL(m_current_block, 0, ());
+      auto const bit = m_current_block & -m_current_block;
+      return bits::PopCount(bit - 1) + m_current_block_index * 64;
```

Comment:
No! CeilLog() requires 6 conditionals, while PopCount() does not use conditionals at all!